cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.operation;
+
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.operation.LongRunningOperation;
+import io.cdap.cdap.common.operation.LongRunningOperationRequest;
+import io.cdap.cdap.proto.id.OperationRunId;
+import io.cdap.cdap.proto.operationrun.OperationError;
+import io.cdap.cdap.proto.operationrun.OperationType;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link OperationRunner} to run an operation in the same service.
+ */
+public class InMemoryOperationRunner<T> extends AbstractOperationRunner<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRunner.class);
+
+  private final OperationStatePublisher statePublisher;
+  private final Lock lock;
+  private OperationDriver<T> driver;
+
+  /**
+   * Default constructor.
+   *
+   * @param operations A map of {@link OperationType} to {@link LongRunningOperation} injected
+   *     via {@link com.google.inject.multibindings.MapBinder}
+   * @param statePublisher Publishes the current operation state.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
```

Comment:
we shouldn't need to setup this mapping with guice bindings because we will never be binding the same operation type to different classes in different scenarios. The extra indirection just makes it harder to tell what is happening. It would be more clear to have a static mapping directly in the code, directly in the abstract class.

Reply:
I also think the class naming is confusing because it mixes terminology that usually means something else in CDAP. 

The current OperationRunner would be better named the LongRunningOperationService, as it looks like this is what will be called from the HTTP handler to actually kick off operations requested from the rest API. The LROService interface may have an in-memory implementation and a task based implementation.

The current LongRunningOperation interface is more like a LongRunningOperationRunner, as it doesn't define *what* an operation is, but *how* to execute an operation. That interface would have one implementation for each OperationType (scm bulk pull, scm bulk push, etc).

Reply:
Remove `public` from `@Inject` constructor.